### PR TITLE
Passw0rd Dev to master

### PR DESCRIPTION
**Pull request by BuddahLD**

### CODE REVIEW

**Review by Scratch-net — DISMISSED** [review]

File: build.gradle [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+            // Http client
+            fuel      : '1.15.1',
+
+            // Kotlin
+            kotlin     : '1.3.11',
+            coroutines : '1.0.1',
+
+            // Gradle
+            gardle     : '3.2.1',
```
Comment (SanjoDeundiak) [inline build.gradle]: Is it correct?

File: passw0rd-protos/src/main/java/com/virgilsecurity/passw0rd/protobuf/proto/passw0rd.proto [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+option java_package = "com.virgilsecurity.passw0rd.protobuf.build";
+option java_outer_classname = "Passw0rdProtos";
+
+message DatabaseRecord {
+    uint32 version = 1;
+    bytes record = 2;
+}
+
+message EnrollmentRequest{
```
Comment (SanjoDeundiak) [inline passw0rd-protos/src/main/java/com/virgilsecurity/passw0rd/protobuf/proto/passw0rd.proto]: Add space

File: passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/client/HttpClientProtobuf.kt [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+import com.github.kittinunf.fuel.Fuel
+import com.google.protobuf.Message
+import com.google.protobuf.Parser
+import com.virgilsecurity.passw0rd.data.ProtocolException
+import com.virgilsecurity.passw0rd.protobuf.build.Passw0rdProtos
+
+/**
+ * . _  _
+ * .| || | _
+ * -| || || |   Created by:
```
Comment (SanjoDeundiak) [inline passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/client/HttpClientProtobuf.kt]: Remove such comments. Copyright above is enough

File: passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/Protocol.kt [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                it.version to it.record.toByteArray()
+            }
+        } catch (e: InvalidProtocolBufferException) {
+            throw InvalidProtobufTypeException()
+        }
+
+        if (pheClients[version] == null)
+            throw NoKeysFoundException("Unable to find keys corresponding to record's version $version.")
+
+        val request = pheClients[version]!!.createVerifyPasswordRequest(password.toByteArray(), record)
```
Comment (SanjoDeundiak) [inline passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/Protocol.kt]: Use of force unwrap is not recommended

Comment (SanjoDeundiak) [reply passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/Protocol.kt]: Try to store pheClients[version] in separate variable and check it for null

File: passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/ProtocolContext.kt [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+                versionedUpdateToken = Passw0rdProtos.VersionedUpdateToken
+                        .newBuilder()
+                        .setVersion(tokenVersion)
+                        .setUpdateToken(ByteString.copyFrom(content))
+                        .build()
+            }
+
+            return ProtocolContext(appToken,
+                                                                        pheClients,
```
Comment (SanjoDeundiak) [inline passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/ProtocolContext.kt]: Fi
   xfor
mat TINg

**Review by SanjoDeundiak — CHANGES_REQUESTED** [review]

File: passw0rd-protos/src/main/java/com/virgilsecurity/passw0rd/protobuf/proto/passw0rd.proto (hunk 2 above)
Comment (BavovnaIsTaken) [reply passw0rd-protos/src/main/java/com/virgilsecurity/passw0rd/protobuf/proto/passw0rd.proto]: Done. (Forgot to format after copying from https://github.com/passw0rd/sdk-go/blob/master/passw0rd.proto)

File: build.gradle (hunk 1 above)
Comment (BavovnaIsTaken) [reply build.gradle]: If you talking about versions difference - yes, it's ok. Formatting is ok too.

File: passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/Protocol.kt (hunk 4 above)
Comment (BavovnaIsTaken) [reply passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/Protocol.kt]: good point

File: build.gradle (hunk 1 above)
Comment (BavovnaIsTaken) [reply build.gradle]: ok, got it

**Review by SanjoDeundiak — DISMISSED** [review]

**Review by SanjoDeundiak — DISMISSED** [review]

**Review by SanjoDeundiak — APPROVED** [review]

**Pull request merged**